File: src/simulation/gplatespython/rift_pathfinder.py
"""Rift pathfinding with A* search, zigzag generation, and craton avoidance."""

import heapq
import math
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class GridCell:
    """A cell in the discretized plate grid."""

    row: int
    col: int
    lon: float
    lat: float
    is_plate: bool = False
    is_ocean_edge: bool = False
    is_craton: bool = False
    g: float = float("inf")
    h: float = 0.0
    f: float = float("inf")
    parent: Optional["GridCell"] = None
    _dist_to_centroid: float = 0.0
    _move_cost: float = 1.0


@dataclass
class RiftPath:
    """Result of rift pathfinding."""

    coordinates: list[list[float]]
    start_point: list[float]
    end_point: list[float]
    total_length_km: float
    zigzag_amplitude_deg: float
    num_zigzags: int


def _haversine_km(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    """Distance between two lon/lat points in kilometers."""
    R = 6371.0
    dlat = math.radians(lat2 - lat1)
    dlon = math.radians(lon2 - lon1)
    a = (
        math.sin(dlat / 2) ** 2
        + math.cos(math.radians(lat1))
        * math.cos(math.radians(lat2))
        * math.sin(dlon / 2) ** 2
    )
    return R * 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))


def _point_in_polygon(
    lon: float, lat: float, polygon: list[list[float]], tolerance: float = 0.1
) -> bool:
    """Ray-casting point-in-polygon test with boundary tolerance."""
    n = len(polygon)
    inside = False
    j = n - 1
    for i in range(n):
        xi, yi = polygon[i][0], polygon[i][1]
        xj, yj = polygon[j][0], polygon[j][1]
        if ((yi > lat) != (yj > lat)) and (
            lon < (xj - xi) * (lat - yi) / (yj - yi + 1e-10) + xi
        ):
            inside = not inside
        j = i

    if not inside:
        for px, py in polygon:
            if abs(px - lon) < tolerance and abs(py - lat) < tolerance:
                inside = True
                break

    return inside


def _build_grid(
    plate_coords: list[list[float]],
    craton_polygons: list[list[list[float]]],
    resolution_deg: float = 1.0,
) -> tuple[list[list[GridCell]], float, float, float, float]:
    """Build a discretized grid over the plate with craton markers."""
    lons = [c[0] for c in plate_coords]
    lats = [c[1] for c in plate_coords]
    min_lon, max_lon = min(lons), max(lons)
    min_lat, max_lat = min(lats), max(lats)

    margin = resolution_deg * 2
    min_lon -= margin
    max_lon += margin
    min_lat -= margin
    max_lat += margin

    rows = max(1, int((max_lat - min_lat) / resolution_deg) + 1)
    cols = max(1, int((max_lon - min_lon) / resolution_deg) + 1)

    grid: list[list[GridCell]] = []
    for r in range(rows):
        row_cells: list[GridCell] = []
        for c in range(cols):
            lon = min_lon + c * resolution_deg
            lat = max_lat - r * resolution_deg
            in_plate = _point_in_polygon(lon, lat, plate_coords)
            in_craton = any(
                _point_in_polygon(lon, lat, craton) for craton in craton_polygons
            )
            cell = GridCell(
                row=r,
                col=c,
                lon=lon,
                lat=lat,
                is_plate=in_plate,
                is_ocean_edge=False,
                is_craton=in_craton and in_plate,
            )
            row_cells.append(cell)
        grid.append(row_cells)

    _mark_ocean_edges(grid, rows, cols, plate_coords, resolution_deg)

    return grid, min_lon, min_lat, max_lon, max_lat


def _mark_ocean_edges(
    grid: list[list[GridCell]],
    rows: int,
    cols: int,
    plate_coords: list[list[float]],
    resolution_deg: float,
) -> None:
    """Mark cells adjacent to non-plate area as ocean edges."""
    for r in range(rows):
        for c in range(cols):
            cell = grid[r][c]
            if not cell.is_plate:
                continue

            is_edge = False
            for dr, dc in [
                (-1, 0),
                (1, 0),
                (0, -1),
                (0, 1),
                (-1, -1),
                (-1, 1),
                (1, -1),
                (1, 1),
            ]:
                nr, nc = r + dr, c + dc
                if nr < 0 or nr >= rows or nc < 0 or nc >= cols:
                    is_edge = True
                    break
                if not grid[nr][nc].is_plate:
                    is_edge = True
                    break

            if is_edge:
                cell.is_ocean_edge = True


def _find_ocean_edge_cells(
    grid: list[list[GridCell]],
    rows: int,
    cols: int,
) -> list[GridCell]:
    """Return all ocean edge cells sorted by plate boundary position."""
    edges = []
    for r in range(rows):
        for c in range(cols):
            if grid[r][c].is_ocean_edge:
                edges.append(grid[r][c])
    return edges


def _select_rift_endpoints(
    edge_cells: list[GridCell],
    plate_coords: list[list[float]],
) -> tuple[GridCell, GridCell]:
    """Select two ocean-edge points far apart for rift start/end."""
    if len(edge_cells) < 2:
        raise ValueError("Need at least 2 ocean edge cells to create a rift")

    centroid_lon = sum(c[0] for c in plate_coords) / len(plate_coords)
    centroid_lat = sum(c[1] for c in plate_coords) / len(plate_coords)

    for cell in edge_cells:
        cell._dist_to_centroid = _haversine_km(
            cell.lon, cell.lat, centroid_lon, centroid_lat
        )

    edge_cells.sort(key=lambda c: c._dist_to_centroid, reverse=True)

    best_pair = (edge_cells[0], edge_cells[1])
    best_dist = _haversine_km(
        edge_cells[0].lon,
        edge_cells[0].lat,
        edge_cells[1].lon,
        edge_cells[1].lat,
    )

    for i in range(min(len(edge_cells), 50)):
        for j in range(i + 1, min(len(edge_cells), 50)):
            d = _haversine_km(
                edge_cells[i].lon,
                edge_cells[i].lat,
                edge_cells[j].lon,
                edge_cells[j].lat,
            )
            if d > best_dist:
                best_dist = d
                best_pair = (edge_cells[i], edge_cells[j])

    return best_pair


def _heuristic(a: GridCell, b: GridCell) -> float:
    """Euclidean heuristic in degrees (scaled for A*)."""
    dlon = a.lon - b.lon
    dlat = a.lat - b.lat
    return math.sqrt(dlon * dlon + dlat * dlat)


def _get_neighbors(
    cell: GridCell,
    grid: list[list[GridCell]],
    rows: int,
    cols: int,
    goal: GridCell,
) -> list[GridCell]:
    """Get valid neighbors for A* expansion."""
    neighbors = []
    for dr, dc in [
        (-1, 0),
        (1, 0),
        (0, -1),
        (0, 1),
        (-1, -1),
        (-1, 1),
        (1, -1),
        (1, 1),
    ]:
        nr, nc = cell.row + dr, cell.col + dc
        if 0 <= nr < rows and 0 <= nc < cols:
            neighbor = grid[nr][nc]
            if neighbor.is_plate and not neighbor.is_craton:
                cost = 1.0
                if dr != 0 and dc != 0:
                    cost = 1.414

                dlon = neighbor.lon - goal.lon
                dlat = neighbor.lat - goal.lat
                dist_to_goal = math.sqrt(dlon * dlon + dlat * dlat)

                if dist_to_goal < 5.0:
                    cost *= 0.5

                neighbor._move_cost = cost
                neighbors.append(neighbor)
    return neighbors


def _astar(
    start: GridCell,
    goal: GridCell,
    grid: list[list[GridCell]],
    rows: int,
    cols: int,
) -> Optional[list[GridCell]]:
    """A* pathfinding from start to goal avoiding cratons."""
    for r in range(rows):
        for c in range(cols):
            grid[r][c].g = float("inf")
            grid[r][c].f = float("inf")
            grid[r][c].parent = None

    start.g = 0
    start.h = _heuristic(start, goal)
    start.f = start.h

    open_set: list[tuple[float, int, GridCell]] = [(start.f, 0, start)]
    counter = 1
    closed_set: set[tuple[int, int]] = set()

    while open_set:
        _, _, current = heapq.heappop(open_set)

        if (current.row, current.col) == (goal.row, goal.col):
            path = []
            node: Optional[GridCell] = current
            while node is not None:
                path.append(node)
                node = node.parent
            path.reverse()
            return path

        if (current.row, current.col) in closed_set:
            continue
        closed_set.add((current.row, current.col))

        for neighbor in _get_neighbors(current, grid, rows, cols, goal):
            if (neighbor.row, neighbor.col) in closed_set:
                continue

            move_cost = getattr(neighbor, "_move_cost", 1.0)
            tentative_g = current.g + move_cost

            if tentative_g < neighbor.g:
                neighbor.parent = current
                neighbor.g = tentative_g
                neighbor.h = _heuristic(neighbor, goal)
                neighbor.f = neighbor.g + neighbor.h
                heapq.heappush(open_set, (neighbor.f, counter, neighbor))
                counter += 1

    return None


def _generate_zigzag(
    base_path: list[GridCell],
    grid: list[list[GridCell]],
    rows: int,
    cols: int,
    amplitude_cells: int = 2,
    zigzag_interval: int = 4,
) -> list[GridCell]:
    """Add zigzag deviations to the base path while avoiding cratons."""
    if len(base_path) < 3:
        return base_path

    zigzagged = [base_path[0]]

    for i in range(1, len(base_path) - 1):
        if i % zigzag_interval == 0:
            direction = 1 if (i // zigzag_interval) % 2 == 0 else -1

            prev_lon = zigzagged[-1].lon
            prev_lat = zigzagged[-1].lat

            best_offset: Optional[GridCell] = None
            best_score = float("inf")

            for offset in range(1, amplitude_cells + 1):
                for dr, dc in [
                    (0, 1),
                    (0, -1),
                    (1, 0),
                    (-1, 0),
                    (1, 1),
                    (-1, -1),
                    (1, -1),
                    (-1, 1),
                ]:
                    nr = base_path[i].row + dr * offset * direction
                    nc = base_path[i].col + dc * offset * direction

                    if 0 <= nr < rows and 0 <= nc < cols:
                        candidate = grid[nr][nc]
                        if candidate.is_plate and not candidate.is_craton:
                            dist_from_base = _haversine_km(
                                candidate.lon,
                                candidate.lat,
                                base_path[i].lon,
                                base_path[i].lat,
                            )
                            dist_to_next = _haversine_km(
                                candidate.lon,
                                candidate.lat,
                                prev_lon,
                                prev_lat,
                            )
                            score = dist_from_base * 0.3 + dist_to_next * 0.7

                            if score < best_score:
                                best_score = score
                                best_offset = candidate

            if best_offset is not None:
                zigzagged.append(best_offset)
                prev_lon = best_offset.lon
                prev_lat = best_offset.lat
            else:
                zigzagged.append(base_path[i])
        else:
            zigzagged.append(base_path[i])

    zigzagged.append(base_path[-1])
    return zigzagged


def _smooth_path(
    path: list[GridCell],
    grid: list[list[GridCell]],
    rows: int,
    cols: int,
    iterations: int = 3,
) -> list[GridCell]:
    """Smooth the path by averaging adjacent points while staying valid."""
    if len(path) < 3:
        return path

    smoothed = list(path)

    for _ in range(iterations):
        new_path = [smoothed[0]]
        for i in range(1, len(smoothed) - 1):
            prev = smoothed[i - 1]
            curr = smoothed[i]
            next_ = smoothed[i + 1]

            avg_lon = (prev.lon + curr.lon + next_.lon) / 3.0
            avg_lat = (prev.lat + curr.lat + next_.lat) / 3.0

            closest = curr
            min_dist = float("inf")
            for dr in range(-1, 2):
                for dc in range(-1, 2):
                    nr = curr.row + dr
                    nc = curr.col + dc
                    if 0 <= nr < rows and 0 <= nc < cols:
                        candidate = grid[nr][nc]
                        if candidate.is_plate and not candidate.is_craton:
                            d = _haversine_km(
                                candidate.lon, candidate.lat, avg_lon, avg_lat
                            )
                            if d < min_dist:
                                min_dist = d
                                closest = candidate
            new_path.append(closest)
        new_path.append(smoothed[-1])
        smoothed = new_path

    return smoothed


def generate_rift_path(
    plate_coords: list[list[float]],
    craton_polygons: list[list[list[float]]],
    resolution_deg: float = 1.0,
    zigzag_amplitude: float = 2.0,
    zigzag_interval: int = 4,
) -> RiftPath:
    """Generate a rift path across a plate from ocean edge to ocean edge.

    Args:
        plate_coords: Polygon coordinates of the plate [[lon, lat], ...]
        craton_polygons: List of craton polygons to avoid
        resolution_deg: Grid cell size in degrees
        zigzag_amplitude: Zigzag deviation in degrees
        zigzag_interval: How many cells between zigzag turns

    Returns:
        RiftPath with coordinates and metadata

    Raises:
        ValueError: If pathfinding fails
    """
    grid, min_lon, min_lat, max_lon, max_lat = _build_grid(
        plate_coords, craton_polygons, resolution_deg
    )

    rows = len(grid)
    cols = len(grid[0]) if rows > 0 else 0

    if rows == 0 or cols == 0:
        raise ValueError("Grid is empty - check plate coordinates")

    edge_cells = _find_ocean_edge_cells(grid, rows, cols)
    if len(edge_cells) < 2:
        raise ValueError(f"Need at least 2 ocean edge cells, found {len(edge_cells)}")

    start, goal = _select_rift_endpoints(edge_cells, plate_coords)

    base_path = _astar(start, goal, grid, rows, cols)
    if base_path is None:
        raise ValueError(
            "No valid path found between ocean edges - "
            "plate may be fully surrounded by cratons"
        )

    amplitude_cells = max(1, int(zigzag_amplitude / resolution_deg))
    zigzagged = _generate_zigzag(
        base_path,
        grid,
        rows,
        cols,
        amplitude_cells=amplitude_cells,
        zigzag_interval=zigzag_interval,
    )

    smoothed = _smooth_path(zigzagged, grid, rows, cols)

    coordinates = [[cell.lon, cell.lat] for cell in smoothed]

    total_length = 0.0
    for i in range(1, len(coordinates)):
        total_length += _haversine_km(
            coordinates[i - 1][0],
            coordinates[i - 1][1],
            coordinates[i][0],
            coordinates[i][1],
        )

    num_zigzags = len(zigzagged) // max(1, zigzag_interval)

    return RiftPath(
        coordinates=coordinates,
        start_point=[start.lon, start.lat],
        end_point=[goal.lon, goal.lat],
        total_length_km=total_length,
        zigzag_amplitude_deg=zigzag_amplitude,
        num_zigzags=num_zigzags,
    )

File: src/simulation/riftPathfinder.js
/**
 * Rift pathfinding with A* search, zigzag generation, and craton avoidance.
 * Ported from Python rift_pathfinder.py for browser use.
 */

class GridCell {
  constructor(row, col, lon, lat, isPlate = false, isOceanEdge = false, isCraton = false) {
    this.row = row
    this.col = col
    this.lon = lon
    this.lat = lat
    this.isPlate = isPlate
    this.isOceanEdge = isOceanEdge
    this.isCraton = isCraton
    this.g = Infinity
    this.h = 0
    this.f = Infinity
    this.parent = null
    this._distToCentroid = 0
    this._moveCost = 1.0
  }
}

function haversineKm(lon1, lat1, lon2, lat2) {
  const R = 6371.0
  const dlat = ((lat2 - lat1) * Math.PI) / 180
  const dlon = ((lon2 - lon1) * Math.PI) / 180
  const a =
    Math.sin(dlat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dlon / 2) ** 2
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a))
}

function pointInPolygon(lon, lat, polygon, tolerance = 0.1) {
  const n = polygon.length
  let inside = false
  let j = n - 1
  for (let i = 0; i < n; i++) {
    const xi = polygon[i][0], yi = polygon[i][1]
    const xj = polygon[j][0], yj = polygon[j][1]
    if ((yi > lat) !== (yj > lat) && lon < ((xj - xi) * (lat - yi)) / (yj - yi + 1e-10) + xi) {
      inside = !inside
    }
    j = i
  }
  if (!inside) {
    for (const [px, py] of polygon) {
      if (Math.abs(px - lon) < tolerance && Math.abs(py - lat) < tolerance) {
        inside = true
        break
      }
    }
  }
  return inside
}

function buildGrid(plateCoords, cratonPolygons, resolutionDeg = 1.0) {
  const lons = plateCoords.map(c => c[0])
  const lats = plateCoords.map(c => c[1])
  let minLon = Math.min(...lons), maxLon = Math.max(...lons)
  let minLat = Math.min(...lats), maxLat = Math.max(...lats)

  const margin = resolutionDeg * 2
  minLon -= margin; maxLon += margin
  minLat -= margin; maxLat += margin

  const rows = Math.max(1, Math.floor((maxLat - minLat) / resolutionDeg) + 1)
  const cols = Math.max(1, Math.floor((maxLon - minLon) / resolutionDeg) + 1)

  const grid = []
  for (let r = 0; r < rows; r++) {
    const rowCells = []
    for (let c = 0; c < cols; c++) {
      const lon = minLon + c * resolutionDeg
      const lat = maxLat - r * resolutionDeg
      const inPlate = pointInPolygon(lon, lat, plateCoords)
      const inCraton = cratonPolygons.some(cr => pointInPolygon(lon, lat, cr))
      rowCells.push(new GridCell(r, c, lon, lat, inPlate, false, inCraton && inPlate))
    }
    grid.push(rowCells)
  }

  markOceanEdges(grid, rows, cols)
  return { grid, rows, cols, minLon, minLat, maxLon, maxLat }
}

function markOceanEdges(grid, rows, cols) {
  const dirs = [[-1,0],[1,0],[0,-1],[0,1],[-1,-1],[-1,1],[1,-1],[1,1]]
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      const cell = grid[r][c]
      if (!cell.isPlate) continue
      let isEdge = false
      for (const [dr, dc] of dirs) {
        const nr = r + dr, nc = c + dc
        if (nr < 0 || nr >= rows || nc < 0 || nc >= cols || !grid[nr][nc].isPlate) {
          isEdge = true
          break
        }
      }
      if (isEdge) cell.isOceanEdge = true
    }
  }
}

function selectRiftEndpoints(edgeCells, plateCoords) {
  if (edgeCells.length < 2) throw new Error('Need at least 2 ocean edge cells')

  const centroidLon = plateCoords.reduce((s, c) => s + c[0], 0) / plateCoords.length
  const centroidLat = plateCoords.reduce((s, c) => s + c[1], 0) / plateCoords.length

  for (const cell of edgeCells) {
    cell._distToCentroid = haversineKm(cell.lon, cell.lat, centroidLon, centroidLat)
  }

  edgeCells.sort((a, b) => b._distToCentroid - a._distToCentroid)

  let bestPair = [edgeCells[0], edgeCells[1]]
  let bestDist = haversineKm(edgeCells[0].lon, edgeCells[0].lat, edgeCells[1].lon, edgeCells[1].lat)

  const limit = Math.min(edgeCells.length, 50)
  for (let i = 0; i < limit; i++) {
    for (let j = i + 1; j < limit; j++) {
      const d = haversineKm(edgeCells[i].lon, edgeCells[i].lat, edgeCells[j].lon, edgeCells[j].lat)
      if (d > bestDist) {
        bestDist = d
        bestPair = [edgeCells[i], edgeCells[j]]
      }
    }
  }
  return bestPair
}

function heuristic(a, b) {
  const dlon = a.lon - b.lon, dlat = a.lat - b.lat
  return Math.sqrt(dlon * dlon + dlat * dlat)
}

function getNeighbors(cell, grid, rows, cols, goal) {
  const neighbors = []
  const dirs = [[-1,0],[1,0],[0,-1],[0,1],[-1,-1],[-1,1],[1,-1],[1,1]]
  for (const [dr, dc] of dirs) {
    const nr = cell.row + dr, nc = cell.col + dc
    if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
      const neighbor = grid[nr][nc]
      if (neighbor.isPlate && !neighbor.isCraton) {
        let cost = (dr !== 0 && dc !== 0) ? 1.414 : 1.0
        const distToGoal = Math.sqrt((neighbor.lon - goal.lon) ** 2 + (neighbor.lat - goal.lat) ** 2)
        if (distToGoal < 5.0) cost *= 0.5
        neighbor._moveCost = cost
        neighbors.push(neighbor)
      }
    }
  }
  return neighbors
}

function astar(start, goal, grid, rows, cols) {
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      grid[r][c].g = Infinity
      grid[r][c].f = Infinity
      grid[r][c].parent = null
    }
  }

  start.g = 0
  start.h = heuristic(start, goal)
  start.f = start.h

  const openSet = [{ f: start.f, counter: 0, cell: start }]
  let counter = 1
  const closedSet = new Set()

  while (openSet.length > 0) {
    openSet.sort((a, b) => a.f - b.f || a.counter - b.counter)
    const current = openSet.shift().cell

    if (current.row === goal.row && current.col === goal.col) {
      const path = []
      let node = current
      while (node) {
        path.push(node)
        node = node.parent
      }
      return path.reverse()
    }

    const key = `${current.row},${current.col}`
    if (closedSet.has(key)) continue
    closedSet.add(key)

    for (const neighbor of getNeighbors(current, grid, rows, cols, goal)) {
      if (closedSet.has(`${neighbor.row},${neighbor.col}`)) continue
      const tentativeG = current.g + neighbor._moveCost
      if (tentativeG < neighbor.g) {
        neighbor.parent = current
        neighbor.g = tentativeG
        neighbor.h = heuristic(neighbor, goal)
        neighbor.f = neighbor.g + neighbor.h
        openSet.push({ f: neighbor.f, counter: counter++, cell: neighbor })
      }
    }
  }
  return null
}

function generateZigzag(basePath, grid, rows, cols, amplitudeCells = 2, zigzagInterval = 4) {
  if (basePath.length < 3) return basePath

  const zigzagged = [basePath[0]]
  const dirs = [[0,1],[0,-1],[1,0],[-1,0],[1,1],[-1,-1],[1,-1],[-1,1]]

  for (let i = 1; i < basePath.length - 1; i++) {
    if (i % zigzagInterval === 0) {
      const direction = (Math.floor(i / zigzagInterval) % 2 === 0) ? 1 : -1
      let prevLon = zigzagged[zigzagged.length - 1].lon
      let prevLat = zigzagged[zigzagged.length - 1].lat

      let bestOffset = null
      let bestScore = Infinity

      for (let offset = 1; offset <= amplitudeCells; offset++) {
        for (const [dr, dc] of dirs) {
          const nr = basePath[i].row + dr * offset * direction
          const nc = basePath[i].col + dc * offset * direction
          if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            const candidate = grid[nr][nc]
            if (candidate.isPlate && !candidate.isCraton) {
              const distFromBase = haversineKm(candidate.lon, candidate.lat, basePath[i].lon, basePath[i].lat)
              const distToNext = haversineKm(candidate.lon, candidate.lat, prevLon, prevLat)
              const score = distFromBase * 0.3 + distToNext * 0.7
              if (score < bestScore) {
                bestScore = score
                bestOffset = candidate
              }
            }
          }
        }
      }

      if (bestOffset) {
        zigzagged.push(bestOffset)
        prevLon = bestOffset.lon
        prevLat = bestOffset.lat
      } else {
        zigzagged.push(basePath[i])
      }
    } else {
      zigzagged.push(basePath[i])
    }
  }

  zigzagged.push(basePath[basePath.length - 1])
  return zigzagged
}

function smoothPath(path, grid, rows, cols, iterations = 3) {
  if (path.length < 3) return path
  let smoothed = [...path]

  for (let iter = 0; iter < iterations; iter++) {
    const newPath = [smoothed[0]]
    for (let i = 1; i < smoothed.length - 1; i++) {
      const prev = smoothed[i - 1], curr = smoothed[i], next = smoothed[i + 1]
      const avgLon = (prev.lon + curr.lon + next.lon) / 3
      const avgLat = (prev.lat + curr.lat + next.lat) / 3

      let closest = curr, minDist = Infinity
      for (let dr = -1; dr <= 1; dr++) {
        for (let dc = -1; dc <= 1; dc++) {
          const nr = curr.row + dr, nc = curr.col + dc
          if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            const candidate = grid[nr][nc]
            if (candidate.isPlate && !candidate.isCraton) {
              const d = haversineKm(candidate.lon, candidate.lat, avgLon, avgLat)
              if (d < minDist) { minDist = d; closest = candidate }
            }
          }
        }
      }
      newPath.push(closest)
    }
    newPath.push(smoothed[smoothed.length - 1])
    smoothed = newPath
  }
  return smoothed
}

function closestPointOnSegment(px, py, ax, ay, bx, by) {
  const abx = bx - ax
  const aby = by - ay
  const apx = px - ax
  const apy = py - ay
  const abLenSq = abx * abx + aby * aby

  if (abLenSq < 1e-12) {
    return { lon: ax, lat: ay }
  }

  const t = Math.max(0, Math.min(1, (apx * abx + apy * aby) / abLenSq))
  return {
    lon: ax + t * abx,
    lat: ay + t * aby
  }
}

function closestPointOnPolygonBoundary(lon, lat, polygon) {
  if (!polygon || polygon.length < 2) {
    return { lon, lat }
  }

  let best = { lon: polygon[0][0], lat: polygon[0][1] }
  let bestDistSq = Infinity

  const lastIndex = polygon.length - 1
  const isClosed =
    Math.abs(polygon[0][0] - polygon[lastIndex][0]) < 1e-9 &&
    Math.abs(polygon[0][1] - polygon[lastIndex][1]) < 1e-9

  const segmentCount = isClosed ? polygon.length - 1 : polygon.length
  for (let i = 0; i < segmentCount; i++) {
    const a = polygon[i]
    const b = polygon[(i + 1) % polygon.length]
    const candidate = closestPointOnSegment(lon, lat, a[0], a[1], b[0], b[1])
    const dx = candidate.lon - lon
    const dy = candidate.lat - lat
    const d2 = dx * dx + dy * dy

    if (d2 < bestDistSq) {
      bestDistSq = d2
      best = candidate
    }
  }

  return best
}

/**
 * Generate a rift path across a plate from ocean edge to ocean edge.
 *
 * @param {number[][]} plateCoords - [[lon, lat], ...]
 * @param {number[][][]} cratonPolygons - [[[lon, lat], ...], ...]
 * @param {object} options
 * @returns {number[][]} Rift coordinates as [[lon, lat], ...]
 */
export function generateRiftPath(plateCoords, cratonPolygons, options = {}) {
  const resolutionDeg = options.resolutionDeg ?? 1.0
  const zigzagAmplitude = options.zigzagAmplitude ?? 2.0
  const zigzagInterval = options.zigzagInterval ?? 4

  const { grid, rows, cols } = buildGrid(plateCoords, cratonPolygons, resolutionDeg)

  const edgeCells = []
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (grid[r][c].isOceanEdge) edgeCells.push(grid[r][c])
    }
  }

  if (edgeCells.length < 2) {
    throw new Error(`Need at least 2 ocean edge cells, found ${edgeCells.length}`)
  }

  const [start, goal] = selectRiftEndpoints(edgeCells, plateCoords)
  const basePath = astar(start, goal, grid, rows, cols)

  if (!basePath) {
    throw new Error('No valid path found between ocean edges')
  }

  const amplitudeCells = Math.max(1, Math.floor(zigzagAmplitude / resolutionDeg))
  const zigzagged = generateZigzag(basePath, grid, rows, cols, amplitudeCells, zigzagInterval)
  const smoothed = smoothPath(zigzagged, grid, rows, cols)

  const coordinates = smoothed.map(cell => [cell.lon, cell.lat])

  // Enforce rift endpoints on the continent frontier (polygon boundary).
  if (coordinates.length >= 2) {
    const start = coordinates[0]
    const end = coordinates[coordinates.length - 1]
    const snappedStart = closestPointOnPolygonBoundary(start[0], start[1], plateCoords)
    const snappedEnd = closestPointOnPolygonBoundary(end[0], end[1], plateCoords)
    coordinates[0] = [snappedStart.lon, snappedStart.lat]
    coordinates[coordinates.length - 1] = [snappedEnd.lon, snappedEnd.lat]
  }

  return coordinates
}
